Bug 546115: show editor guiding text

marcushoepfner opened the pull request:
First draft of showing some guiding text in the editor area.

Open: where to put the actual code, at the moment it is in StackRenderer.
Also at the moment there is a leak with on of the images. But let's first discuss whether it should be done like this or the whether the image and text should come from a new extension point.

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/META-INF/MANIFEST.MF:
```diff
@@ -27,7 +27,8 @@ Require-Bundle: org.eclipse.e4.ui.workbench;bundle-version="0.9.0",
  org.eclipse.equinox.preferences;bundle-version="3.5.200",
- org.eclipse.core.runtime;bundle-version="3.17.100"
+ org.eclipse.core.runtime;bundle-version="3.17.100",
+ org.eclipse.ui;bundle-version="[3.201.0,4.0.0)"
```
You can't add `org.eclipse.ui` dependency here

marcushoepfner replied:
removed

vogella commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -623,6 +637,66 @@ public Object createWidget(MUIElement element, Object parent) {
+
+		tabFolder.setOnboarding(container);
+
+		IPageService service = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getService(IPageService.class);
```
I think it is easier to use the IEventBroker here and you don't require the org.eclispe.ui dependency in this  case

marcushoepfner replied:
done, but I don't get informed with the EventBroker. Maybe subscription is too late. But nevertheless I guess I should get informed when the perspective is changed?!

vogella commented:
Hey, great to see that you started to work on this. I think this will make a huge different is UX for the IDE

vogella commented:
Currently not at my PC but you should be able to use annotations to do the event registration.  IIRC StackRender uses this already for other events, try to find methods which are using @UIEventTopic.

marcushoepfner commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -209,6 +216,13 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
+	@Inject
+	@Optional
+	void subscribePerspectiveChanged(
+			@UIEventTopic(UIEvents.UILifeCycle.PERSPECTIVE_OPENED) org.osgi.service.event.Event event) {
```
I tried this one. It works only if the perspective is opened via "Open perspective dialog". It does not work if I just click a perspective button (e.g. java or debug) in the toolbar. The even is not raised then.
There are other events in UIEvents.Perspective (e.g. TOPIC_ALL). But is is also not raised.

Basically I would need a notification on startup and on perspective change. But without an osgi dependency to another plugin.

Any ideas?

laeubi replied:
> It works only if the perspective is opened via "Open perspective dialog". It does not work if I just click a perspective button (e.g. java or debug) in the toolbar. The even is not raised then.

Probably this should be fixed then?

Beside that, you can use `UIEvents.ElementContainer.TOPIC_SELECTEDELEMENT` and then do:

```
Object element = event.getProperty(EventTags.ELEMENT);
if(element instanceof MPerspectiveStack) {
	MPerspective perspective = ((MPerspectiveStack)element).getSelectedElement();
	if(perspective != null) {
                          //do something with it...
	}
}
```

marcushoepfner replied:
Ok, I have created a new ui event called PERSPECTIVE_SWITCHED.
With that it works.

marcushoepfner commented:
I'd say with the last commit it is the first version which works.

But I'm not sure whether all the code is placed correctly.

And - shame on me - I still don't know whether I have to increase plugin version or not? Sorry!

mickaelistria commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/internal/workbench/swt/OnboardingCTabFolder.java:
```diff
@@ -0,0 +1,55 @@
+package org.eclipse.e4.ui.internal.workbench.swt;
```
Please add a copyright header (copy and adapt from other files)

marcushoepfner replied:
Done

mickaelistria commented on bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/workbench/UIEvents.java:
```diff
@@ -310,6 +310,13 @@ public interface UILifeCycle {
+		/**
+		 * Sent when a perspective is switched
+		 *
+		 * @since ???
```
Must be the x.y of next release for this bundle. As it's a new API, bundle version must become `x.y+1.0`.
So please change bundle version to `1.14.0` in MANIFEST.MF, and set this to `@since 1.14`

marcushoepfner replied:
Done

BeckerWdf reviewed (commented):
Looks good to me for a first implementation.
The question is now how can a perspective contribute specific content to this. 
Like the top n most useful shortcuts and a dedicated image.

BeckerWdf commented:
Just for reference. This is the bugzilla bug for this: https://bugs.eclipse.org/bugs/show_bug.cgi?id=546115

marcushoepfner commented:
> Looks good to me for a first implementation. The question is now how can a perspective contribute specific content to this. Like the top n most useful shortcuts and a dedicated image.

That would be the next step. E.g. extend the Perspective Extension Point with some fields or create a new extension point. That was my idea at least.

BeckerWdf commented:
> > Looks good to me for a first implementation. The question is now how can a perspective contribute specific content to this. Like the top n most useful shortcuts and a dedicated image.
> 
> That would be the next step. E.g. extend the Perspective Extension Point with some fields or create a new extension point. That was my idea at least.

That sounds good.

mickaelistria reviewed (approved):
I'm fine with this change, however, I think it's worth leaving some more time for @vogella to have a new look at it before merging (don't forget to use "Squash and merge"!)

vogella commented:
/rebase

vogella commented:
Currently fighting corvid so please do not wait for me. Would be great to have this functionality.

mickaelistria commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/.settings/.api_filters:
```diff
@@ -8,4 +8,12 @@
             </message_arguments>
         </filter>
     </resource>
+    <resource path="src/org/eclipse/e4/ui/internal/workbench/swt/OnboardingCTabFolder.java" type="org.eclipse.e4.ui.internal.workbench.swt.OnboardingCTabFolder">
```
Why is it necessary? This is internal code so API tools should not run here.

marcushoepfner replied:
I got an API warning in the IDE. I must not extend CTabFolder (it has @noextend annotation).
Eclipse proposed a quick fix which added the entry in the .api_filter file.

The file was already there and it contained an entry already. So the quick fix sounded it reasonable.

Do you think I should remove this entry?

laeubi replied:
@marcushoepfner you probably could *wrap* a CTab control instead of *extending* it and using PaintListener/ControlListenr/... instead? I uses this to show a nice image instead of the gray background for empty editor area.

marcushoepfner replied:
Where do you use this? Do you have an example?
Simplicity is always welcome :)

laeubi replied:
```
CTabFolder cTabFolder = ....

cTabFolder.addPaintListener(new PaintListener() {

	@Override
	public void paintControl(PaintEvent e) {

		if(cTabFolder.getItemCount() == 0) {		
			GC gc = e.gc;
			Color background = gc.getBackground();
			Color foreground = gc.getForeground();
			Point size = control.getSize();
			try {
				paintBackground(gc, background, size, bgImage);
			} finally {
				gc.setBackground(background);
				gc.setForeground(foreground);
			}
		}
  	}
}
```

iloveeclipse replied:
I would argue here following: the warning is not there to be ignored, but to re-think the approach. Instead of **extending** CTabFolder (which is bad) you could use ordinary CTabFolder with a listener attached.

So please undo the API change and please try to use default CTabFolder with resize listener.

marcushoepfner commented:
Shall merge this?
There is are build failures but I don't understand them.

mickaelistria commented:
> Shall merge this? There is are build failures but I don't understand them.

If there are test failures that are not understood, then no, it shouldn't be merged.
Moreover the project is currently code-frozen for some days to produce the M1 build, merge shouldn't happen until promotion of M1 is complete. Some notifications are sent on the eclipse-dev@eclipse.org mailing-list.

marcushoepfner commented:
> > Shall merge this? There is are build failures but I don't understand them.
> 
> If there are test failures that are not understood, then no, it shouldn't be merged. Moreover the project is currently code-frozen for some days to produce the M1 build, merge shouldn't happen until promotion of M1 is complete. Some notifications are sent on the [eclipse-dev@eclipse.org](mailto:eclipse-dev@eclipse.org) mailing-list.

Sorry, I know. What I actually wanted to say:
1. are we through with the review part?
2. Does someone understand the build error. As usual I'm struggling. I wish we had https://plugins.jenkins.io/build-failure-analyzer/.

merks commented:
It's always so hard to find these darned API errors, but I think this is the key part:
```
2 API ERRORS
* Marker [on: /org.eclipse.e4.ui.workbench/META-INF/MANIFEST.MF, id: 686, type: org.eclipse.pde.api.tools.version_numbering, attributes: [apiMarkerID: 4, charEnd: 145, charStart: 129, description: - The field org.eclipse.e4.ui.workbench.UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED in an interface that is intended to be implemented or extended has been added
, lineNumber: 4, message: The major version should be incremented in version 1.14.0, since API breakage occurred since version 1.13.100, messagearguments: 1.14.0#1.13.100, problemid: 923795461, severity: 2, sourceId: API Tools, version: 2.0.0.qualifier], created: 6/29/22, 8:58 AM]
* Marker [on: /org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/workbench/UIEvents.java, id: 685, type: org.eclipse.pde.api.tools.compatibility, attributes: [apiMarkerID: 2, charEnd: 9823, charStart: 9803, lineNumber: 318, message: The field org.eclipse.e4.ui.workbench.UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED in an interface that is intended to be implemented or extended has been added, messagearguments: org.eclipse.e4.ui.workbench.UIEvents.UILifeCycle#PERSPECTIVE_SWITCHED, org.eclipse.jdt.internal.core.JavaModelManager.handleId: =org.eclipse.e4.ui.workbench/src<org.eclipse.e4.ui.workbench{UIEvents.java[UIEvents[UILifeCycle^PERSPECTIVE_SWITCHED, problemTypeName: org.eclipse.e4.ui.workbench.UIEvents$UILifeCycle, problemid: 403767336, severity: 2, sourceId: API Tools], created: 6/29/22, 8:58 AM]
```
It thinks you did something that's API breaking.

marcushoepfner commented:
> It's always so hard to find these darned API errors, but I think this is the key part:
> 
> ```
> 2 API ERRORS
> * Marker [on: /org.eclipse.e4.ui.workbench/META-INF/MANIFEST.MF, id: 686, type: org.eclipse.pde.api.tools.version_numbering, attributes: [apiMarkerID: 4, charEnd: 145, charStart: 129, description: - The field org.eclipse.e4.ui.workbench.UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED in an interface that is intended to be implemented or extended has been added
> , lineNumber: 4, message: The major version should be incremented in version 1.14.0, since API breakage occurred since version 1.13.100, messagearguments: 1.14.0#1.13.100, problemid: 923795461, severity: 2, sourceId: API Tools, version: 2.0.0.qualifier], created: 6/29/22, 8:58 AM]
> * Marker [on: /org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/workbench/UIEvents.java, id: 685, type: org.eclipse.pde.api.tools.compatibility, attributes: [apiMarkerID: 2, charEnd: 9823, charStart: 9803, lineNumber: 318, message: The field org.eclipse.e4.ui.workbench.UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED in an interface that is intended to be implemented or extended has been added, messagearguments: org.eclipse.e4.ui.workbench.UIEvents.UILifeCycle#PERSPECTIVE_SWITCHED, org.eclipse.jdt.internal.core.JavaModelManager.handleId: =org.eclipse.e4.ui.workbench/src<org.eclipse.e4.ui.workbench{UIEvents.java[UIEvents[UILifeCycle^PERSPECTIVE_SWITCHED, problemTypeName: org.eclipse.e4.ui.workbench.UIEvents$UILifeCycle, problemid: 403767336, severity: 2, sourceId: API Tools], created: 6/29/22, 8:58 AM]
> ```
> 
> It thinks you did something that's API breaking.

I have added UIEvents.PERSPECTIVE_SWITCHED and increased minor.
Shall I really increase the major? Do we do such things? Major some implies incompatibility. At least to me.

mickaelistria commented:
The class has a `@noextend` flag in documentation, so it is OK to add fields. However, API Tools should take that into account and not fail. @vik-chand Am I right? Is this issue known or should we report it to PDE?
@marcushoepfner If you have setup API Tools in your IDE with a correct baseline (latest release), then you should see this error in your IDE and have a quick-fix to create an API Analysis filter for that case. This will edit a file that you should commit together with your change.

laeubi commented:
> API Tools should take that into account and not fail. @vik-chand Am I right?

I think we have talked about it some where else, this is even an interface and adding a new field to an interface should always be compatible.

marcushoepfner commented:
> The class has a `@noextend` flag in documentation, so it is OK to add fields. However, API Tools should take that into account and not fail. @vik-chand Am I right? Is this issue known or should we report it to PDE? @marcushoepfner If you have setup API Tools in your IDE with a correct baseline (latest release), then you should see this error in your IDE and have a quick-fix to create an API Analysis filter for that case. This will edit a file that you should commit together with your change.

How do I setup the API baseline?

laeubi commented:
@mickaelistria the interface `org.eclipse.e4.ui.workbench.UIEvents` is marked as noextend/implement, but the embedded type `org.eclipse.e4.ui.workbench.UIEvents.UILifeCycle` is **not**. This might be not intentional, but API tools is not to blame here.

marcushoepfner commented:
> > The class has a `@noextend` flag in documentation, so it is OK to add fields. However, API Tools should take that into account and not fail. @vik-chand Am I right? Is this issue known or should we report it to PDE? @marcushoepfner If you have setup API Tools in your IDE with a correct baseline (latest release), then you should see this error in your IDE and have a quick-fix to create an API Analysis filter for that case. This will edit a file that you should commit together with your change.
> 
> How do I setup the API baseline?

Got help already. API baseline is setup. I have uploaded a new commit with .api_filters file.

mickaelistria commented:
https://github.com/eclipse-platform/eclipse.platform.ui/pull/219 is related to the API check.

marcushoepfner commented:
@iloveeclipse , your request for change (in the early beginning of this change) is blocking merge. Could you kindly re-review. Dependency to o.e.ui has been removed. thx

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench/META-INF/MANIFEST.MF:
```diff
@@ -55,3 +55,9 @@ Require-Capability: osgi.extender;
   filter:="(objectClass=org.osgi.service.event.EventAdmin)";
   effective:="active"
 Automatic-Module-Name: org.eclipse.e4.ui.workbench
+Service-Component: OSGI-INF/org.eclipse.e4.ui.internal.LocaleChangeServiceContextFunction.xml,
```
Please undo this change, you should see warnings in IDE about duplicated headers.

marcushoepfner replied:
Done

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/internal/workbench/swt/OnboardingCTabFolder.java:
```diff
@@ -0,0 +1,81 @@
+import org.eclipse.swt.widgets.Listener;
+
+/**
+ * A CTabFolder which shows a set {@link Composite} in case there are no tabs.
```
All what this (illegally) extended class does could be done by a legal SWT listener added to a regular CTabFolder. Please try this approach.

marcushoepfner replied:
Done

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -209,6 +218,32 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
+	private void setOnboarding(MPerspective perspective) {
+		welcomeText.getDisplay().asyncExec(() -> {
+			Image image = getImageFromURI(perspective.getIconURI());
+			Image greyImage = new Image(welcomeText.getDisplay(), image, SWT.IMAGE_GRAY);
```
Both images are leaked. Please make sure they are properly disposed.

marcushoepfner replied:
refer to the other comment about new state

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -623,6 +667,38 @@ public Object createWidget(MUIElement element, Object parent) {
+
+		tabFolder.addDisposeListener(e -> {
+			if (!color.isDisposed()) {
+				color.dispose();
```
Colors don't need to be disposed anymore.

marcushoepfner replied:
removed it

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -623,6 +667,38 @@ public Object createWidget(MUIElement element, Object parent) {
+				color.dispose();
+			}
+			if (welcomeImage.getImage() != null && !welcomeImage.getImage().isDisposed()) {
+				welcomeImage.dispose();
```
This looks strange, we've created two images but dispose one. Easier is to add dispose listener to the control on which you supply images at the moment you create them.

marcushoepfner replied:
Done! But I dont really like it. I keep the onboarding as state now.
together with welcomeImage and welcomeText this is a lot new state. I'm not a fan of state.

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -623,6 +667,38 @@ public Object createWidget(MUIElement element, Object parent) {
+		Composite content = WidgetFactory.composite(SWT.NONE).supplyLayoutData(gridData::create)
+				.layout(GridLayoutFactory.swtDefaults().create()).create(container);
+
+		Color color = container.getDisplay().getSystemColor(SWT.COLOR_WIDGET_DISABLED_FOREGROUND);
```
I'm not a user of Dark mode, but does it work there too?

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -623,6 +667,38 @@ public Object createWidget(MUIElement element, Object parent) {
 	}
 
+	/**
+	 * @param tabFolder
```
Please remove useless javadoc

marcushoepfner replied:
Deon

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -209,6 +218,32 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
+	}
+
+	private void setOnboarding(MPerspective perspective) {
+		welcomeText.getDisplay().asyncExec(() -> {
```
This will fail with NPE if there was no "org.eclipse.e4.primaryDataStack" tag at the "createWidget()" time.

marcushoepfner replied:
fixed

iloveeclipse reviewed (dismissed):
@marcushoepfner: Beside fixing multiple issues below, could you please **squash** all commits to one and always do an amend and force push? 

Searching for changes in ~15 commits in IDE is not a fun, and review of this change in github doesn't make sense too, it is not a trivial patch.

mickaelistria commented:
FWIW, now that the inner types in UIEvents are marked as `@noimplements` then the API filter may not be necesary anymore.

vogella commented:
@marcushoepfner are you current available to look into the feedback from @iloveeclipse?

marcushoepfner commented:
> @marcushoepfner are you current available to look into the feedback from @iloveeclipse?

@vogella Unfortunately not before mid of September.

marcushoepfner commented:
I've uploaded a new version with less files changed. Reason: I used the standard CTabFolder with listeners.
Unfortunately image and text is not removed if the first file is opened.
Did not find a listener in CTabFolder for that.
No clue how that can be implemented.

Also the close is currently buggy I think since I get informed when the user attempts to close the last tab, not when the tab is already closed. Some code might interfere and cancel the close (e.g. the save dialog).

Any ideas? Thanks

mickaelistria commented:
Another approach would be that if no editor is shown, then we open a default editor which has the right guiding content (some would argue that what's a good welcome page should do, but it's unfortunately not what the current welcome page is doing; maybe it's just an opportunity to make the welcome page more useful to users).
And whenever another editor is open, then we close the "guide".

This would be way less intrusive and allow to capitalize on existing bits instead of creating something new.

BeckerWdf commented:
> Another approach would be that if no editor is shown, then we open a default editor which has the right guiding content (some would argue that what's a good welcome page should do, but it's unfortunately not what the current welcome page is doing; maybe it's just an opportunity to make the welcome page more useful to users). And whenever another editor is open, then we close the "guide".
> 
> This would be way less intrusive and allow to capitalize on existing bits instead of creating something new.

But the idea of this to provide something less intrusive to the user. Just like VScode does it when no file is open.

mickaelistria commented:
> But the idea of this to provide something less intrusive to the user. Just like VScode does it when no file is open.

the only difference would be a tabitem visible on the top bar of editor stack whem empty. FWIW, this is how we implemented that a while ago before I worked for Red Hat: https://documentation.bonitasoft.com/bonita/7.9/bonita-studio-download-installation (see the welcome page). The users and other stakeholders were pretty happy with it and it didn't require deep changes, it was plain workbench usage in a dedicated plugin. The content can be even simpler and stick to the 3 links shown in the screenshot on bugzilla.
IMO the proposed approach is overkill for the functional requirement. I'd rather seeing Platform trying to keep things simple instead of growing capabilities so it becomes harder and harder to maintain them all.

BeckerWdf commented:
> the only difference would be a tabitem visible on the top bar of editor stack whem empty. 

No it's not. The guiding text would be very passive. You can move around an editor tab etc. This is about subtile and intrusive hints to the user (hence the gray water mark style color). Think of a user opening an Eclipse based tool for the first time and seeing a very empty screen. I would say this:
![image](https://user-images.githubusercontent.com/28338612/186852618-304477b7-c555-4000-a3ea-d08ff2917c93.png)
is from a UX perspective something completely different than this:
![image](https://user-images.githubusercontent.com/28338612/186852740-859f3fff-7067-42d9-9f05-90dd3a5c5086.png)
> IMO the proposed approach is overkill for the functional requirement. 

I disagree.
> I'd rather seeing Platform trying to keep things simple instead of growing capabilities so it becomes harder and harder to maintain them all.

Platform's goal should best UX and not easiest implementation for platform devs. An in VsCode this has proven to be a good approach from UX point of view. And I would think that VsCode most likely has more UX experts than platform has.

laeubi commented:
> Unfortunately image and text is not removed if the first file is opened.
> Did find a listener in CTabFolder for that.
> No clue how that can be implemented.

Whats wrong with the suggested Paint listener?

vogella commented:
I also think that this functionality would hugely improve UX

BeckerWdf commented:
> I also think that this functionality would hugely improve UX

Just do be sure: Which approach do you mean with "this"?

vogella commented:
> Just do be sure: Which approach do you mean with "this"?

I think any guidince is better than at the moment, I like the second screenshot which IMHO is very close to VScode.

mickaelistria commented:
I wish that we can think more and more by ourselves about what would really help Eclipse Platform's users and how we can already help without creating new relatively expensive frameworks to mimic competitors when we already have some existing ones that are supposed to already cover the user stories. As long as we're just following, we're not really innovating.
I also wish we would always place first users at the center of such UX discussion instead of blindly mimicking VSCode.
There are a lot of good things in VSCode, but not all of them are IMO good to copy in Eclipse IDE. Eclipse IDE is on some aspects superior to what VSCode can offer and we should capitalize on those, especially if it brings more maintenance sustainability.

IMO, in this case, just showing the welcome page (and improving it if need be) when there is no editor open would lead to a better result than this proposal, as it would allow RCP providers to keep their customization, would allow some more dynamic content (can even be showing an external web page)...

laeubi commented:
> when there is no editor open

**Just from a user POV:** I almost never have "no editor open" because I usually use my IDE to code some stuff ;-)

So "Recent files" or even a "Quick search" (like google start page) would be much more useful than any "Welcome", or "Shortcut" or even "Website" stuff... I almost never can think about that I did more than closing the welcome page (what ever product) ...

laeubi commented:
If one want tome really cool stuff, then something like the JIRA Dashboard would be something one can think of so if I close all editors, I can see what are my next issues to fix:

https://www.atlassian.com/blog/jira-software/5-steps-to-build-a-killer-dashboard

BeckerWdf commented:
> If one want tome really cool stuff, then something like the JIRA Dashboard would be something one can think of so if I close all editors, I can see what are my next issues to fix:
> 
> https://www.atlassian.com/blog/jira-software/5-steps-to-build-a-killer-dashboard

This is not about using empty (wasted) space but about onboarding new users. So I don't see how the dashboard example related to this.

vogella commented:
F*ck, the default "update" button in Github does a merge. Sorry, can you update this commit to do a rebase, I don't think I can do this. I will now extract the event for perspective switching into a new commit, this should be non-controversial and reduce the size of this change.

marcushoepfner commented on bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/workbench/UIEvents.java:
```diff
@@ -316,6 +316,13 @@ public interface UILifeCycle {
+		 *
+		 * @since 1.13
+		 */
+		String PERSPECTIVE_SWITCHED = TOPIC + TOPIC_SEP + "perspSwitched"; //$NON-NLS-1$
```
as soon as #394  is merged, file can be removed

marcushoepfner commented on bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/internal/workbench/PartServiceImpl.java:
```diff
@@ -627,6 +627,7 @@ public void switchPerspective(MPerspective perspective) {
 
 				modelService.bringToTop(target);
 				activate(target, true, false);
+				UIEvents.publishEvent(UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED, perspective);
```
as soon as #394 is merged, file can be removed

marcushoepfner commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -209,6 +217,42 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
+				welcomeImage.dispose();
+			});
+
+			String text = NLS.bind("Welcome to {0}", perspective.getLabel()); //$NON-NLS-1$
```
as the commit msg says, it is planned to e.g. outsource text, image and maybe actions to an extension point.
Can this text stay NON-NLS until then?

Otherwise this would be the first externalized string in this plugin and would lead to Message.java, messages file and new package for both.

marcushoepfner reviewed (commented):
The current implementation with a PaintListener works AFAICS.
WDYT?

Are there still leaks?
Pls. review.

vogella commented:
@marcushoepfner please rebase onto master

marcushoepfner commented:
> @marcushoepfner please rebase onto master

Done

vogella commented:
FYI During testing I encountered https://github.com/eclipse-platform/eclipse.platform.ui/pull/434, I assume this is not directly related to this change.

marcushoepfner commented:
As often, I'm struggling with the build errors. Does not build successful, but I cannot see any meaningful error description.
Could someone help pls?

iloveeclipse commented:
Click on details.
Go to https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/27/#showFailuresLink
See 12 failimg tests

vogella commented:
@marcushoepfner LGTM, planning to merge once / if the validation build is green. Would be nice to add actions to the editor area in the future.

vogella commented:
@marcushoepfner any idea why the tests are failing?

marcushoepfner commented:
> @marcushoepfner any idea why the tests are failing?

No, not really. Currently struggling to set them in my local IDE. But I'm working on it.

marcushoepfner commented:
> > @marcushoepfner any idea why the tests are failing?
> 
> No, not really. Currently struggling to set them in my local IDE. But I'm working on it.

I cannot find out why these tests are failing.
Setting them up locally is a mess. Mockito, hamcrest, bytebuddy, o.e.tests.harness. jface tests...
I have installed/imported so many plugins and it is still not working.

laeubi commented:
You can run the maven build with `-Dtest=<your test to investigate>` and use remote debugging:
https://tycho.eclipseprojects.io/doc/latest/tycho-surefire-plugin/test-mojo.html#debugPort

vogella commented:
> > > @marcushoepfner any idea why the tests are failing?
> > 
> > 
> > No, not really. Currently struggling to set them in my local IDE. But I'm working on it.
> 
> I cannot find out why these tests are failing. Setting them up locally is a mess. Mockito, hamcrest, bytebuddy, o.e.tests.harness. jface tests... I have installed/imported so many plugins and it is still not working.

I documented the setup here a while ago https://www.vogella.com/tutorials/EclipsePlatformDevelopment/article.html#running-the-eclipse-platform-unit-tests

BeckerWdf commented:
> > > > @marcushoepfner any idea why the tests are failing?
> > > 
> > > 
> > > No, not really. Currently struggling to set them in my local IDE. But I'm working on it.
> > 
> > 
> > I cannot find out why these tests are failing. Setting them up locally is a mess. Mockito, hamcrest, bytebuddy, o.e.tests.harness. jface tests... I have installed/imported so many plugins and it is still not working.
> 
> I documented the setup here a while ago https://www.vogella.com/tutorials/EclipsePlatformDevelopment/article.html#running-the-eclipse-platform-unit-tests

Can we somehow put that on the github org so that this is easy to find?

laeubi commented:
> Can we somehow put that on the github org so that this is easy to find?

https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/CONTRIBUTING.md
You should first sync this with:
https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/CONTRIBUTING

iloveeclipse commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -209,6 +217,42 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
+			return;
+		}
+
+		onBoarding.getDisplay().asyncExec(() -> {
```
Why is this code runs in asyncExec? Please note. `asyncExec()` might run at any time in the future, even if the perspective already closed, but there are no checks at all inside the body if the code still should run.

marcushoepfner replied:
I don't remember. Was a back and forth during development.
I removed it. Let's see whether the tests are working now.

marcushoepfner commented:
> > > > @marcushoepfner any idea why the tests are failing?
> > > 
> > > 
> > > No, not really. Currently struggling to set them in my local IDE. But I'm working on it.
> > 
> > 
> > I cannot find out why these tests are failing. Setting them up locally is a mess. Mockito, hamcrest, bytebuddy, o.e.tests.harness. jface tests... I have installed/imported so many plugins and it is still not working.
> 
> I documented the setup here a while ago https://www.vogella.com/tutorials/EclipsePlatformDevelopment/article.html#running-the-eclipse-platform-unit-tests

I followed the instructions and installed mockito and hamcrest from https://download.eclipse.org/tools/orbit/downloads/drops/R20221123021534/ into my 2022-12 dev ide. (Target is 2023-03)

But I still get two errors in the manifest of o.e.ui.tests:
![Screenshot 2022-12-22 at 09 40 56](https://user-images.githubusercontent.com/29301444/209093665-f5e23450-9f24-4d77-a5af-9cc562d0067c.png)
So I cannot run the locally in dev ide. Any Idea?

akurtakov commented:
The tutorial is outdated. Please install org.eclipse.test feature from current stream https://download.eclipse.org/eclipse/updates/4.27-I-builds/ . That would give you both. Orbit is less and less used for platform development.

iloveeclipse commented:
Also there is https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md

iloveeclipse commented:
Tests are still failing:
```
[Test Result](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/) (13 failures / +2)
[org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testInitialSelectionWithTabbedProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testInitialSelectionWithTabbedProperties/)
[org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug268676HidingPinnedTargetPart](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug268676HidingPinnedTargetPart/)
[org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514TabbedProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514TabbedProperties/)
[org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514TabbedProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514TabbedProperties_2/)
[org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testPageDispose](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testPageDispose/)
[org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514NormalProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514NormalProperties/)
[org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514NormalProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514NormalProperties_2/)
[org.eclipse.ui.tests.commands.Bug417762Test.testSuggestionUseParameterizedCommandExecuteWithChecks](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testSuggestionUseParameterizedCommandExecuteWithChecks/)
[org.eclipse.ui.tests.commands.Bug417762Test.testSuggestionUseExecuteCommand](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testSuggestionUseExecuteCommand/)
[org.eclipse.ui.tests.commands.Bug417762Test.testAsReported](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testAsReported/)
[org.eclipse.ui.tests.commands.Bug417762Test.testSuggestionUseExecuteCommandInContext](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testSuggestionUseExecuteCommandInContext/)
[org.eclipse.ui.tests.menus.MenuBaseTests.testBasicMenuPopulation](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.menus/MenuBaseTests/testBasicMenuPopulation/)
[org.eclipse.ui.tests.menus.MenuPopulationTest.testViewPopulation](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.menus/MenuPopulationTest/testViewPopulation/)
```

marcushoepfner commented:
> The tutorial is outdated. Please install org.eclipse.test feature from current stream https://download.eclipse.org/eclipse/updates/4.27-I-builds/ . That would give you both. Orbit is less and less used for platform development.

In the target IDE I guess, correct?

marcushoepfner commented:
> Tests are still failing:
> 
> ```
> [Test Result](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/) (13 failures / +2)
> [org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testInitialSelectionWithTabbedProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testInitialSelectionWithTabbedProperties/)
> [org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug268676HidingPinnedTargetPart](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug268676HidingPinnedTargetPart/)
> [org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514TabbedProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514TabbedProperties/)
> [org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514TabbedProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514TabbedProperties_2/)
> [org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testPageDispose](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testPageDispose/)
> [org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514NormalProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514NormalProperties/)
> [org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.testBug278514NormalProperties](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.propertysheet/MultiInstancePropertySheetTest/testBug278514NormalProperties_2/)
> [org.eclipse.ui.tests.commands.Bug417762Test.testSuggestionUseParameterizedCommandExecuteWithChecks](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testSuggestionUseParameterizedCommandExecuteWithChecks/)
> [org.eclipse.ui.tests.commands.Bug417762Test.testSuggestionUseExecuteCommand](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testSuggestionUseExecuteCommand/)
> [org.eclipse.ui.tests.commands.Bug417762Test.testAsReported](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testAsReported/)
> [org.eclipse.ui.tests.commands.Bug417762Test.testSuggestionUseExecuteCommandInContext](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.commands/Bug417762Test/testSuggestionUseExecuteCommandInContext/)
> [org.eclipse.ui.tests.menus.MenuBaseTests.testBasicMenuPopulation](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.menus/MenuBaseTests/testBasicMenuPopulation/)
> [org.eclipse.ui.tests.menus.MenuPopulationTest.testViewPopulation](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-78/35/testReport/junit/org.eclipse.ui.tests.menus/MenuPopulationTest/testViewPopulation/)
> ```

Locally in Eclipse when running as PDE test I'm getting this for MultiInstancePropertySheetTest:
```
java.lang.IllegalStateException: Workbench has not been created yet.
	at org.eclipse.ui.PlatformUI.getWorkbench(PlatformUI.java:117)
	at org.eclipse.ui.tests.harness.util.CloseTestWindowsRule.addWindowListener(CloseTestWindowsRule.java:56)
	at org.eclipse.ui.tests.harness.util.CloseTestWindowsRule.before(CloseTestWindowsRule.java:40)
	at org.eclipse.ui.tests.harness.util.UITestCase.setUp(UITestCase.java:181)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:74)
	at org.eclipse.pde.internal.junit.runtime.CoreTestApplication.start(CoreTestApplication.java:28)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)

java.lang.NullPointerException: Cannot invoke "org.eclipse.ui.IWorkbenchPage.resetPerspective()" because "this.activePage" is null
	at org.eclipse.ui.tests.propertysheet.MultiInstancePropertySheetTest.doTearDown(MultiInstancePropertySheetTest.java:120)
	at org.eclipse.ui.tests.harness.util.UITestCase.tearDown(UITestCase.java:214)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunAfters.invokeMethod(RunAfters.java:46)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:74)
	at org.eclipse.pde.internal.junit.runtime.CoreTestApplication.start(CoreTestApplication.java:28)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
```

While e.g. MenuBaseTests is green.

So I cannot really reproduce what is happening on Jenkins. Are these errors really due to my commit?

BeckerWdf commented:
> Are these errors really due to my commit?

Yes.

BeckerWdf commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -209,6 +217,40 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
+			return;
+		}
+
+		Image image = getImageFromURI(perspective.getIconURI());
```
This should be:

```suggestion
		Image image = getImageFromURI(perspective.getIconURI());
		if (image.isDisposed()) {
			return;
		}
```

This at least fixed the failing test locally on my computer

BeckerWdf commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -209,6 +217,40 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
+			welcomeImage.dispose();
+		});
+
+		String text = NLS.bind("Welcome to {0}", perspective.getLabel()); //$NON-NLS-1$
```
```suggestion
		String text = NLS.bind("Welcome to the \"{0}\" perspective", perspective.getLabel()); //$NON-NLS-1$
```

marcushoepfner replied:
thx

BeckerWdf commented on bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java:
```diff
@@ -623,6 +673,44 @@ public Object createWidget(MUIElement element, Object parent) {
+		tabFolder.addPaintListener(e -> setOnboardingControlSize(tabFolder, onBoarding));
+
+		tabFolder.addDisposeListener(e -> {
+			if (welcomeImage.getImage() != null && !welcomeImage.getImage().isDisposed()) {
```
```suggestion
			if (welcomeImage != null && !welcomeImage.isDisposed() && welcomeImage.getImage() != null && !welcomeImage.getImage().isDisposed()) {
```

marcushoepfner replied:
thx

marcushoepfner committed:
Show editor guiding text #78

Adds an "OnboardingComposite" to editor CTabFolder.
The OnboardingComposite contains a greyed image of the current
perspective and a "Welcome to <perspective" text.

A paintlistener controls the size of this OnboardingComposite.
If no editor is open, the OnboardingComposite is shown. As soon as
editor opens it is hidden.

A perspective switch listener changes the image and the label text.

On another change this will be further implemented, e.g. by a new
extension point. In this extension point a larger image could be
defined. And also some more text, e.g. a bullet list with action ("What
can be done in this perspective?").

Signed-off-by: Marcus Hoepfner <marcus.hoepfner@sap.com>

BeckerWdf commented:
> org.eclipse.test

What do you mean with "install"?

BeckerWdf commented:
The final solution should be to have an extension point per perspective what does configure what do show here (a separte icon and which commands to list).
Until this we could bring the first commit into a state so that nothing is visible to the user. With a following commit we could add the extension point and enable this feature.

marcushoepfner commented:
> The final solution should be to have an extension point per perspective what does configure what do show here (a separte icon and which commands to list). Until this we could bring the first commit into a state so that nothing is visible to the user. With a following commit we could add the extension point and enable this feature.

I can do so. Any idea whether just to hide something, or comment code or even remove some of the code? Which way to go?
